lib/12T/magic/ext4mag.py

```diff
@@ -0,0 +1,195 @@
+	combinations can be applied to the following command: 
+	python3 ext4mag.py file_name.mag
+
+	By default, the Makefile included in this directory will run the following format:
+	python3 ext4mag.py -move -ap -ng
+
+	Note: This script is intended to be used inside of the magic cell library directory. If
+	the library directory is moved, please change the directory paths below appropriately to ensure a
+	successful script execution.
+"""
+
+import subprocess
+import traceback
+import shutil
+import sys
+import os
+
+class Ext4Mag():
+	def __init__(self):
+		self.files = os.listdir()
+		self.move_flag = False
+		self.add_params = False
+		self.exng = False 
+		self.enter = True
+		self.cwd = os.getcwd()
+
+		os.chdir('../../char/techfiles/')
+
+		self.pdk_spice = os.getcwd()
+		os.chdir(self.cwd)
+
+		if (len(sys.argv) > 1):
+			self.cmdl_parse(sys.argv)
+
+	def cmdl_parse(self, args):
+		for i in range(len(args)):
+			if '.mag' in args[i]:
+				if self.enter:
+					self.enter = False
+					self.files = []
+				self.files.append(args[i])
+			if '-move' in args[i]:
+				self.move_flag = True
+			if '-ap' in args[i]:
+				self.add_params = True
+			if '-ng' in args[i]:
+				self.exng = True
+				
+	
+	def move_files(self):
+		names = ['.sim', '.spice', '.gds', '.ext', '.cif']
+		files = os.listdir()
+
+		for f in files:
+			for s in names:
+				if s in f and '.magicrc' not in f:
+					if not os.path.exists(s.split(".")[1]):
+						os.mkdir(s.split(".")[1])
+					shutil.move(f'./{f}', f'./{s.split(".")[1]}/{f}')
+	
+	def run_magic(self):
+		for f in self.files:
+			if '.mag' in f and '.magicrc' not in f:
+				print(f'Current magic file: {f}')
+				string='drc off\nextract all\next2sim -y 6\next2spice subcircuit top on\next2spice -F -y 6 -f hspice\next2sim merge aggressive\next2sim -y 6 -R -C -o ' + f.split(".")[0] + '.sim\ncif\ncalma\nquit -noprompt\n'
+
+				try:
+					print(f'Extracting files from {f.split(".")[0]}.')
+					subout = subprocess.run(['magic -dnull -noconsole ' + str(f) +  ' 1> MAGIC.LOG 2>&1'], shell=True, input=string.encode()) 
+				except Exception:
+					print(traceback.print_exc())
+
+	def extract_ngspice(self):
+		for f in self.files:
+			if '.mag' in f and '.magicrc' not in f:
+				print(f'Current magic file for ngspice extraction: {f}')
+				string='drc off\nextract\next2spice scale off\next2spice subcircuit top off\next2spice -F -y 6 -f ngspice\nquit -noprompt\n'
+
+				try:
+					print(f'Extracting files from {f.split(".")[0]}.')
+					subout = subprocess.run(['magic -dnull -noconsole ' + str(f) +  ' 1> MAGIC.LOG 2>&1'], shell=True, input=string.encode()) 
+				except Exception:
+					print(traceback.print_exc())
+		
+		if ( not os.path.exists('ngspice')):
+			os.mkdir('ngspice')
+
+		files = os.listdir()
+
+		for f in files:
+			if ((('.spice' in f) or ('.ext' in f)) and (not os.path.isdir(f))):
+				shutil.move(f'./{f}', f'./ngspice/{f}')
```

Comment:
did you consider https://docs.python.org/3/library/pathlib.html for easier path/file operation?

Reply:
not needed - deleted.